src/ChunkedBase.jl

```diff
@@ -6,103 +6,39 @@ using CodecZlibNG
 using NewlineLexers
 using SentinelArrays.BufferedVectors
 
-const MIN_TASK_SIZE_IN_BYTES = 16 * 1024
+# The means through which the user can provide their own parsing logic.
+include("ParsingContexts.jl")
 
-# populate_result_buffer!(result_buf::AbstractResultBuffer, newlines::AbstractVector{Int32}, parsing_ctx::AbstractParsingContext, comment::Union{Nothing,Vector{UInt8}}=nothing, ::Type{CT}=Tuple{}) where {CT}
-function populate_result_buffer! end
-
-abstract type AbstractParsingContext end
-
-# Synchronization mechanism -- after we lexed all rows, we split them in N tasks and TaskCounter
-# in ChunkingContext will block the io/lexer to overwrite the current chunk unless workers
-# report back N times that they are done with their tasks.
+# A counter-based synchronization primitive that is used to coordinate the parsing/consuming tasks.
 include("TaskCounters.jl")
 using .TaskCounters
```

Comment:
Also maybe not in scope for this week, but I _think_ this entire file can be replaced with a Base.Semaphore!
@nickrobinson251 should be able to explain semaphores in detail if you're not familiar; he and I reviewed those at the beginning of this year.

I might be missing something but it looks to me like that's equivalent to what you've built here.

Reply:
I think at some point I was looking into Semaphores, but decided they weren't the right fit, but don't remember the details. How would that work? When the IO tasks finds newlines and sends out N units of work, do we acquire N from the semafore and let each worker release? How would we wait for the semafore to be 0? How would error handling work with a Semaphore? I don't see how it would be an improvement, but maybe I'm missing something.

Reply:
please open an issue about this, if there isn't one - thanks!